File: solutions_5648941810974720_1/Python/flurm/A_Getting_the_Digits.py
#!/usr/bin/python

import sys, collections, pprint

INPUT = sys.stdin
INPUT = open(sys.argv[1],'r')

def getline():
    return INPUT.readline()[:-1]

class Bunch:
    def __init__( self, **kwds ):
        self.__dict__ = kwds


pause_after_trace = False
def mtrace(*strs):
    # return
    atrace(*strs)

def atrace(*strs):
    f = sys.stderr
    print >> f, '..',
    for str in strs:
        print >> f, str,
    print >> f
    if pause_after_trace:
        response = raw_input('? ')
        if response == 'q':
            sys.exit(1)

def memoize(f):
    memos = {}
    def memoized_f( *args ):
        try:
            result = memos[args]
            mtrace(args, ": got result from memo")
        except KeyError:
            result = f(*args)
            mtrace(args, ": got result by calling")
            memos[args] = result
        return result
    return memoized_f

# ------------------------------------------------------------------------------

english_reps = [ "ZERO", "ONE", "TWO", "THREE", "FOUR", "FIVE", "SIX", "SEVEN", "EIGHT", "NINE" ]
letter_counts_for_digit_ = []
for digit, rep in enumerate(english_reps):
    letter_counts_ = collections.defaultdict(int)
    for letter in rep:
        letter_counts_[letter] += 1
    letter_counts_for_digit_.append(letter_counts_)
# pprint.pprint(letter_counts_for_digit_)

extraction_order = [
    ('Z',0),
    ('W',2),
    ('U',4),
    ('X',6),
    ('G',8),
    #
    ('H',3), # or 'R' or 'T'
    ('F',5),
    ('S',7),
    #
    ('O',1),
    ('I',9),
]

n_cases = int(getline())
mtrace(n_cases,'cases:')
for case_num in range(1,n_cases+1):
    mtrace()
    atrace( 'Case #', case_num )

    letter_count_ = collections.defaultdict(int)
    letters = getline()
    for letter in letters:
        letter_count_[letter] += 1

    digit_count_ = {}

    for (letter, digit) in extraction_order:
        k = letter_count_[letter]
        digit_count_[digit] = k

        for (d_letter, d_count) in letter_counts_for_digit_[digit].items():
            letter_count_[d_letter] -= d_count * k
            assert letter_count_[d_letter] >= 0
    
    assert max(letter_count_.values()) == 0

    phone_num = ''
    for (digit, count) in sorted(digit_count_.items()):
        phone_num += ( str(digit) * count )

    print 'Case #%d: %s' % (case_num, phone_num)
    sys.stdout.flush()

assert INPUT.readline() == ''

# vim: sw=4 ts=4 expandtab

File: solutions_5648941810974720_0/java/RogerB/AGettingTheDigits.java
package Round1B;

import java.util.Scanner;

public class AGettingTheDigits {
    static int[][] digitCounts;
    
    static int[] count(String input) {
        int[] result = new int[26];
        for (int x = 0; x < input.length(); x++) result[input.charAt(x) - 'A']++;
        return result;
    }
    
    static boolean subtract(int[] a, int[] b) {
        boolean isValid = true;
        for (int x = 0; x < 26; x++) {
            a[x] -= b[x];
            if (a[x] < 0) isValid = false;
        }
        return isValid;
    }
    
    static void add(int[] a, int[] b) {
        for (int x = 0; x < 26; x++) a[x] += b[x];
    }
    
    static boolean isZero(int[] count) {
        for (int x = 0; x < 26; x++) if (count[x] != 0) return false;
        return true;
    }
    
    static String getDigits(int[] count, int minFirst) {
        if (isZero(count)) return "";
        for (int guessDigit = minFirst; guessDigit <= 9; guessDigit++) {
            boolean isValid = subtract(count, digitCounts[guessDigit]);
            if (isValid) {
                String result = getDigits(count, guessDigit);
                if (!result.equals("ERROR")) {
                    return guessDigit + result;
                }
            }
            add(count, digitCounts[guessDigit]);
        }
        return "ERROR";
    }
    
    public static void main (String[] args) {
        digitCounts = new int[][] {count("ZERO"), count("ONE"), count("TWO"), count("THREE"), count("FOUR"), count("FIVE"), count("SIX"), count("SEVEN"), count("EIGHT"), count("NINE")};
        Scanner sc = new Scanner(System.in);
        int tests = sc.nextInt();
        for (int t = 1; t <= tests; t++) {
            System.out.printf("Case #%d: %s%n", t, getDigits(count(sc.next()), 0));
        }
        sc.close();
    }
}
